sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TableProvider.java

```diff
@@ -93,4 +93,18 @@ default Transform[] inferPartitioning(CaseInsensitiveStringMap options) {
   default boolean supportsExternalMetadata() {
     return false;
   }
+
+  /**
+   * Returns true if the source supports defining generated columns upon table creation in SQL.
+   * When false: any create/replace table statements with a generated column defined in the table
+   * schema will throw an exception during analysis.
+   *
+   * A generated column is defined with syntax: {@code colName colType GENERATED ALWAYS AS (expr)}
+   * The generation expression is stored in the column metadata with key "generationExpression".
+   *
+   * Override this method to allow defining generated columns in create/replace table statements.
+   */
+  default boolean supportsGeneratedColumnsOnCreation() {
```

Comment:
It's a bit weird to define this API in `TableProvider`, which doesn't even have `createTable` function. I think `TableCatalog` is a better place.

Reply:
then it should be in `Table`? In the ideal ds v2 code path, `TableProvider` is not event used: `TableCatalog` -> `Table` -> `ScanBuilder` ...

Reply:
There is no Table object because we haven't created a table before running `CREATE TABLE`. In other words, we only know the table provider name, and we need to know whether this table provider name supports generated columns.

Reply:
Then I think it's `TableCatalog`'s responsibility to decide if a table supports generated column or not.
1. For general catalogs like HMS, it may need to resolve the table provider name to `TableProvider`. Then it makes sense to add a new API to `TableProvider`.
2. For dedicated catalogs like JDBC catalog, it doesn't even have table provider. It depends on the remote SQL database and `TableCatalog` needs a new API to indicate it.

How about we add the new API to both `TableCatalog` and `TableProvider`?

Reply:
Sounds like it would need to be more like a CatalogCapability? Supports Creating Tables with Generated Columns? 

Although in general i'm not sure how most other engines would be able to support this without supporting Spark Expressions. Do we have a limit on how much of the Spark API is included here?

Reply:
@sunchao `TableCapability` has the same issue described above as including it in `Table`
@RussellSpitzer I think it would be up to the implementing data source to restrict what's supported?

Reply:
This really should be a part of the greater the catalog capabilities since that's where create table is usually going to be invoked. I'm very nervous with saying that it is up to the Datasource to decide what is valid because different engines may decide the same sql means different things and this would require that the Datasource somehow make sure non spark engines can access the same table in the same. We spent a lot of time making an public expression class for the connectors but it feels like that should probably be invoked here as well?

Reply:
I've refactored this PR to instead add `TableCatalogCapability` as suggested. Please let me know what you think.